Doc/tools/extensions/grammar_snippet.py

```diff
@@ -0,0 +1,192 @@
+import re
+
+from docutils import nodes
+from docutils.parsers.rst import directives
+from sphinx import addnodes
+from sphinx.util.docutils import SphinxDirective
+from sphinx.util.nodes import make_id
+
+
+class GrammarSnippetDirective(SphinxDirective):
+    """Transform a grammar-snippet directive to a Sphinx literal_block
+
+    That is, turn something like:
+
+        .. grammar-snippet:: file
+           :group: python-grammar
+
+           file: (NEWLINE | statement)*
+
+    into something similar to Sphinx productionlist, but better suited
+    for our needs:
+    - Instead of `::=`, use a colon, as in `Grammar/python.gram`
+    - Show the listing almost as is, with no auto-aligment.
+      The only special character is the backtick, which marks tokens.
+
+    Unlike Sphinx's productionlist, this directive supports options.
+    The "group" must be given as a named option.
+    The content must be preceded by a blank line (like with most ReST
+    directives).
+    """
+
+    has_content = True
+    option_spec = {
+        'group': directives.unchanged,
+    }
+
+    # We currently ignore arguments.
+    required_arguments = 0
+    optional_arguments = 1
+    final_argument_whitespace = True
+
+    def run(self):
+        return make_snippet(self, self.options, self.content)
+
+
+def make_snippet(directive, options, content):
```

Comment:
Would you consider implementing `CompatProductionList` as a subclass of `GrammarSnippet` and overriding the relevent bits? It feels a bit odd to pass the full `directive` object to a free-standing object, though I understand why it is being done.

Reply:
Conceptually, `CompatProductionList` is not a `GrammarSnippet`, so it feels wrong to subclass. The option/argument handling should be entirely separate, I don't want settings from `GrammarSnippet` to become defaults for `CompatProductionList`.
So I added a common base.